src/api/utils.nim

```diff
@@ -0,0 +1,148 @@
+import 
+    unittest,
+    strutils,
+    options,
+    times
+
+import 
+    jsony
+
+proc parseHook*[T](s: string, i: var int, v: var Option[seq[T]]) =
+    eatSpace(s, i)
+    if i + 3 < s.len and
+            s[i+0] == 'n' and
+            s[i+1] == 'u' and
+            s[i+2] == 'l' and
+            s[i+3] == 'l':
+        i += 4
+        return
+    if s[i] == '[':
+        var v2: seq[T]
+        parseHook(s, i, v2)
+        v = some(v2)
+    else:
+        var v2: T
+        parseHook(s, i, v2)
+        v = some(@[v2])
+
+
+proc dumpHook*(s: var string, v: Option[DateTime]) =
+    s.add("\"" & v.get().format("yyyy-MM-dd'T'hh:mm:ss'.'fff'Z'") & "\"" )
+
+proc dumpHook*(s: var string, v: DateTime) =
+    s.add("\"" & v.format("yyyy-MM-dd'T'hh:mm:ss'.'fff'Z'") & "\"" )
+
+proc parseHook*(s: string, i: var int, v: var DateTime) =
+    ## jsony time convert
+    ## runs through times and tries to parse them
+    var str: string
+    parseHook(s, i, str)
+    var timeFormats = @["yyyy-MM-dd", "yyyy-MM-dd hh:mm:ss", "yyyy-MM-dd'T'hh:mm:ss'.'fff'Z'"]
+    for fmt in timeFormats:
+        try:
+            v = parse(str, fmt)
+            return
+        except:
+            continue
+    raise newException(ValueError, "Invalid date format: " & str)
+
+proc parseHook*(s: string, i: var int, v: var int) =
+    ## attempt to parse Ints
+    var str: string
+    parseHook(s, i, str)
+    v = parseInt(str)
+
+proc parseHook*(s: string, i: var int, v: var float) =
+    ## attempt to parse Floats
+    var str: string
+    parseHook(s, i, str)
+    v = parseFloat(str)
+
+proc parseHook*(s: string, i: var int, v: var Option[bool]) =
+    # attempt to parse Bools
+    var str: string
+    parseHook(s, i, str)
+    v = some(parseBool(str))
+
+proc renameHook*(v: object, fieldName: var string) =
+    # loosely match field  to names
+    # MyField -> myfield
+    # myField -> myFields
+    runnableExamples:
+        type
+            MyTest = object
+                id: string
+                myFancyField: string
+
+        var myJson = """
+        {
+            "Id": "someId",
+            "MyFancyField": "foo"
+        }
+        """
+        let myTest =  myJson.fromJson(MyTest)
+        echo myTest
+
+    # MyField -> myField
+    var tempFieldName = fieldName
+    tempFieldName[0] = tempFieldName[0].toLowerAscii()
+    for x , _  in v.fieldPairs():
+        if tempFieldName == x:
+            fieldName = tempFieldName
+            return
+    # try to match with an upload-> uploads
+    tempFieldName &= "s"
+    for x , _  in v.fieldPairs():
+        if tempFieldName == x:
+            fieldName = tempFieldName
+            return
+
+suite "utility functions":
```

Comment:
This should not be present on release

Reply:
Not sure exactly which part you are referring to 


```nim
    # try to match with an upload-> uploads
    tempFieldName &= "s"
    for x , _  in v.fieldPairs():
        if tempFieldName == x:
            fieldName = tempFieldName
            return
            
```
This part is needed because the response from AWS is XML and the response given is something like 
```xml
<Response>
   <Value>Val</Value>
   <Value>Val</Value>
   <Value>Val</Value>
</Response>
```
and we have models like (which is correct). So we need to coerce it or we accept the name of `value` rather then `values`
```nim
Response = object
   values: Seq[String]
```

Reply:
Also I should note: The XML2Json.nim in this library should be made obsolete in favour of 
https://github.com/treeform/taggy when taggy supports deserialisation of XML to Nim objects

Reply:
Hi Lyndon,
Sorry for not being more specific. It was the `suite "utility functions":`. I'll move it to a test directory, don't worry about it.